### ISSUE TITLE
Interaction between softDelete and populate on get()

### ISSUE THREAD

**Issue by eddyystop** [feathers-plus/feathers-hooks-common#284]
softDelete does an initial `get()` to see if the record is logically deleted. This caused 2 get's to be executed on a get() call which was considered bad in some use cases, see https://github.com/feathers-plus/feathers-hooks-common/issues/163 .

So that initial get was not performed on a get() call. However eventually some issues were reported. Here's a history of possibly related issues:
https://github.com/feathers-plus/feathers-hooks-common/issues/270
https://github.com/feathers-plus/feathers-hooks-common/issues/257
https://github.com/feathers-plus/feathers-hooks-common/issues/238
https://github.com/feathers-plus/feathers-hooks-common/issues/163

There also is likely a bad interaction between softDelete and authentication.

The problem is likely here https://github.com/feathers-plus/feathers-hooks-common/blob/master/lib/services/soft-delete.js#L58 which was introduced by https://github.com/feathers-plus/feathers-hooks-common/issues/163

**Comment by eddyystop** [comment]
Nov 15, 2017
======
ravism [7:05 AM] 
Can someone help me understand when using softDelete and populate, it works just fine on a find(), but does not on a get().

```// schema
include: [
    {
      service: '/users',
      nameAs: 'user',
      parentField: 'user_id',
      childField: '_id',
      query: { $disableSoftDelete: true },
}]
```


This works perfectly fine with find(), and populates the deleted user, but when I try to use get() the populated user is just and empty {} object. 

Also this happen always regardless if user is deleted or not.

If I remove the softDelete hook, this on users that are deleted it errors, and for user that are not deleted it works just fine.

If I called the service directly users.get(id, {query: { $disableSoftDelete: true }})
I get back the deleted user.

To me this seems to be an isolated issue with using populate with softDelete on a get hook.

Any help or understand would be much appreciated. I have been stuck on this all day.

**Trying to add more inform regarding what I am trying do.**
I am calling a service (messages) to get a message that has an user that needs to be populate. If I call messages.find({}) and get all messages, each message populates user just fine, however if I call messages.get(id) and only get one message this is where it fails, and does not populate the user. below is the code in messages for the populate. If I user.get(id) I can get the deleted user just fine.
```exports.before = {
  all: [
    softDelete(),
  ],
}

exports.after = {
  all: [],
  find: [
    populate({ schema }),
    serialize(serializeSchema),
  ],
  get: [
    populate({ schema }),
    serialize(serializeSchema),
  ],
  ...
}

const schema = {
  include: [
    {
      service: '/users',
      nameAs: 'user',
      parentField: 'user_id',
      childField: '_id',
      query: { $disableSoftDelete: true },
    },
  ]
}```

**Comment by travisbmiller** [comment]
@eddyystop I have been playing around with this over the last two week, and what I have found to be true is the following:

If I call a service like
``` app.service('/users').get(id, { provider: undefined })```,
 I am doing this so I can use 
```iff(isProvider('external'), auth.authenticate('jwt'))```

If I check params at users before all hook, provider is set to undefined, however users has a populate, lets say "posts", when the populate calls the post service provider is no longer ```{ provider: undefined }```, in my case it is ```{ provider: 'socketio'}```.  So ```iff(isProvider('external'), auth.authenticate('jwt'))``` is true and it try to authenticate, however there is no passed user so it error with no auth token. 

Is this the right behavior? or should the populate pass the ```{provider:undefined}``` from the first call to users service? If this is the right behavior what would be the right way to handle this or a work around?

Thanks

**Comment by travisbmiller** [comment]
The work around I found, was in the schema to add provider: undefined. Not sure if this is how it should be done or not, but it allow it to work for now.

**Comment by eddyystop** [comment]
There has been at least one more report of such an interaction. The fellow did not respond back with some details I asked for, and I haven't been able to reproduce the problem.

softDelete does a probing `get` before every call, except if that call is itself a `get` (so as not to do 2 `get`s). Using the original get with whatever params it has, as opposed to a more more carefully constructed probing get, seems to be the crux of the issue. https://github.com/feathers-plus/feathers-hooks-common/blob/master/lib/services/soft-delete.js#L58

I have to leave atm. I'll read more deeply when I can.

**Comment by eddyystop** [comment]
Maybe related from Slack:
@eddyystop found it, I followed this guide https://docs.feathersjs.com/guides/auth/recipe.custom-auth-strategy.html and had to refactor it from
```const commonHooks = require('feathers-hooks-common');
const auth = require('@feathersjs/authentication');

module.exports = () =>
  commonHooks.iff(
    commonHooks.every(commonHooks.isProvider('external')),
    commonHooks.iffElse(
      ctx => ctx.params.headers['x-api-key'],
      auth.hooks.authenticate('apiKey'),
      auth.hooks.authenticate(['jwt'])
    )
  );
```
to
```const { iff, iffElse, isProvider, every } = require('feathers-hooks-common');
const auth = require('@feathersjs/authentication');

module.exports = function (options = {}) {
  return async context => {
    iff(
      every(isProvider('external')),
      iffElse(
        context => context.params.headers['x-api-key'],
        auth.hooks.authenticate('apiKey'),
        auth.hooks.authenticate(['jwt'])
      )
    );
  };
};```

eddyystop [10:25 AM] 
@roel Your hook is used like `all: myHook()` while the guide example is used like `all: myHook` (which can be confusing as the common hooks all use a called function syntax). Even then the guide example uses `() =>` instead of `ctx =>`. (edited)
 
+1 on figuring that out. Could you create an issue on the docs? This could be one of the reasons there are reports of a conflict between softDelete and auth.

**Comment by eddyystop** [comment]
`softDelete2` is a noteable improvement over `softDelete` especially with regard to authentication along with its restrictToOwner hook. softDelete2 will be released later this week once its docs are complete.

`fastJoin` is also a noteable improvement over `populate` and its gives you total control over the service calls being made.

I'm going to close this issue. Please feel free to open another one if you find issues with softDelete2, or issues with fastJoin and softDelete2.

Thanks.

**Comment by eddyystop** [comment]
Published as v4.15.0
